File: services/client/src/protocol/channel.go
package protocol

import (
	"encoding/binary"
	"fmt"
	"math"
	"net"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

const (
	typeBytes   = 1
	lengthBytes = 2
)

const (
	startBetsSendingType          = 0x01
	filledBetsType                = 0x02
	finalizeBetsSendingType       = 0x03
	startBetWinnersSendingType    = 0x04
	betWinnerType                 = 0x05
	finalizeBetWinnersSendingType = 0x06
)

type MessageChannel struct {
	connection net.Conn
}

func NewMessageChannel(connection net.Conn) *MessageChannel {
	return &MessageChannel{connection: connection}
}

func (channel *MessageChannel) Send(message OutgoingMessage) error {
	payloadBytes := message.Payload()
	if len(payloadBytes) > math.MaxUint16 {
		return fmt.Errorf(
			"payload of %d bytes does not fit in the %d byte length field",
			len(payloadBytes), lengthBytes,
		)
	}

	frameBytes := make([]byte, 0, typeBytes+lengthBytes+len(payloadBytes))
	frameBytes = append(frameBytes, message.Type())
	frameBytes = binary.BigEndian.AppendUint16(frameBytes, uint16(len(payloadBytes)))
	frameBytes = append(frameBytes, payloadBytes...)

	return safe_socket.SendAll(channel.connection, frameBytes)
}

func (channel *MessageChannel) Receive() (IncomingMessage, error) {
	messageType, err := safe_socket.RecvAll(channel.connection, typeBytes)
	if err != nil {
		return nil, err
	}

	messageLength, err := safe_socket.RecvAll(channel.connection, lengthBytes)
	if err != nil {
		return nil, err
	}

	payloadLength := int(binary.BigEndian.Uint16(messageLength))
	payloadBytes, err := safe_socket.RecvAll(channel.connection, payloadLength)
	if err != nil {
		return nil, err
	}

	return incomingMessageFrom(messageType[0], payloadBytes)
}

func incomingMessageFrom(messageType byte, payloadBytes []byte) (IncomingMessage, error) {
	switch messageType {
	case startBetWinnersSendingType:
		return StartBetWinnersSendingMessageFrom(payloadBytes)
	case betWinnerType:
		return BetWinnerMessageFrom(payloadBytes)
	case finalizeBetWinnersSendingType:
		return FinalizeBetWinnersSendingMessageFrom(payloadBytes)
	default:
		return nil, fmt.Errorf("unknown message type 0x%02x", messageType)
	}
}

File: services/server/src/protocol/channel.py
import safe_socket

from protocol.errors import ProtocolError
from protocol.messages.deserialization.incoming import (
    FinalizeBetsSendingMessage,
)
from protocol.messages.deserialization.incoming import FilledBetsMessage, StartBetsSendingMessage

_TYPE_BYTES = 1
_LENGTH_BYTES = 2

_MESSAGE_BY_TYPE = {
    b"\x01": StartBetsSendingMessage,
    b"\x02": FilledBetsMessage,
    b"\x03": FinalizeBetsSendingMessage,
}


class MessageChannel:
    def __init__(self, client_socket):
        self._socket = client_socket

    def receive(self):
        message_type = safe_socket.recv_all(self._socket, _TYPE_BYTES)
        length = int.from_bytes(
            safe_socket.recv_all(self._socket, _LENGTH_BYTES), byteorder="big"
        )
        payload = safe_socket.recv_all(self._socket, length) if length else b""

        message_class = _MESSAGE_BY_TYPE.get(message_type)
        if message_class is None:
            raise ProtocolError(f"unknown message type 0x{message_type.hex()}")
        return message_class.from_bytes(payload)

    def send(self, message):
        payload = message.payload()
        length = len(payload).to_bytes(_LENGTH_BYTES, byteorder="big")
        total_frame_bytes = message.type() + length + payload
        safe_socket.send_all(self._socket, total_frame_bytes)
